sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Sum.scala

```diff
@@ -56,7 +56,13 @@ case class Sum(child: Expression) extends DeclarativeAggregate with ImplicitCast
     case _ => DoubleType
   }
 
-  private lazy val sumDataType = resultType
+  private lazy val sumDataType = child.dataType match {
+    case LongType => DecimalType.BigIntDecimal
```

Comment:
this reduces the possibilities of overflow, but do other databases have the same behavior? Changing the data type of an expression is a breaking change.

Reply:
this is not changing the result data type of the expression, this is changing only the internal buffer type in roder to let "temporary" overflows to happen without any exception. Please consider the case when you have:
```
Long.MaxValue
100
-1000
```
The result should be `Long.MaxValue - 900`. With this buffer type larger than the returned type, we can overflow temporarily when we add `Long.MaxValue` and `100` and then get back to a valid value when we add `-1000`. So with this change we return the correct value. Other DBs behave in this way too.